### PR TITLE
Change type information for out intent to flow from called function like returns

### PR DESCRIPTION
**Pull request by mppf**
This PR adjusts the handling of `out` formals to Option 2 in issue #16582 :

2. Allow all actual arguments to resolve for an `out` formal but then
   produce a compilation error for an unresolved `init=` or `=` between
   the types if one is not present

One implication of this change is that `out` intent formal types are not
determined by the call site, but rather by the called function. In this
way they are more like `return`. As a result, generic `out` intent formal
arguments need to be initialized according to the rules of split
initialization.

Without this PR, ‘out’ intent formal arguments participate in candidate
selection and disambiguation just `ref` formal arguments. That means that
you can, for example, have multiple overloads of a function using the
`out` intent, for example, `proc foo(out arg: int)` and `proc foo(out
arg: real)`.

At the same time, when adjusting the language to optimize copies with
split-init and copy-elision, we decided to consider `out` intent as a way
to initialize uninitialized memory (rather than simply a documentation-y
way to decorate a `ref` intent). That means that, for example, in

``` chapel
proc returnRecordOut(out arg: MyRecord) { ... }
var x: MyRecord; 
returnRecordOut(x);
```

the `var x` line leaves `x` uninitialized because `x` is initialized by
the function call.

In this way, `out` arguments are similar to `return`. The change
implemented in this PR makes them more similar to `return` in two main
ways:
1.	The type of `out` formal arguments is not considered when
        selecting candidates or doing disambiguation (so calling `foo` if
        it has overloads as above would result in ambiguity)
2.	For generic `out` formal arguments, the type is inferred from the
        called function rather than the call site.

The benefits of this approach are:
* `out` becomes more like a different `return` slot
* conversions are allowed for `out` function calls using `=` or `init=`
  as one might expect/hope. Conversions are already allowed for returns
  so this seems natural. Allowing conversions for `out` won’t carry the
  same risks as implicit conversions for `in` (say) because they don’t
  impact which function is called so don’t bring much potential for
  confusion. 
* it allows us to extend split-init support to type-less `out` argument
  calls:

  ``` chapel
  var x, y, z;
  var res = threeOutFormals(x,y,z)
  ```

  because the type of x/y/z is inferred from the called function. This
  makes it easy to transform code returning a tuple into code returning
  multiple out arguments.

Known drawbacks of this approach are:
* it sometimes comes up that code wishes to choose a function according
  to a type to be returned. This comes up in existing code
  channel.readbits and buffer.copyout. In practice, these existing cases
  can just use `ref` intent under the proposal, but that makes their
  signatures less obvious. (Another way such code can work is to add a
  type argument for the result type but that might make calling these
  functions less easy).
* generic out arguments have a more restricted use since the type must be
  known by the called function and the out variable must be
  split-initialized

Besides the above changes, this PR makes a few other improvements:
 * added a type `dtSplitInitType` to represent an unknown type for a
   variable declared as `var x;`. Such a variable's type must be
   established later. Adjusted `gSplitInit` to have type
   `dtSplitInitType`.
 * adjusted toString(ArgSymbol, bool) to have the bool enable/disable the
   intent as well as the type. This makes sense when the result is
   printed in a quoted expression. Added a toString(Symbol, bool) that
   calls the ArgSymbol or VarSymbol one as needed. Improved
   toString(type) to print out `<any type> ` for dtAny. 
 
Also note:
  * special handling for out intents with runtime types is removed in
    commit d9439fd5e9b40d0455bf1abea71d0242490439aa. This can be removed
    because now the type information flows from the called function to
    the call site (instead of from the call site to the called function)
    which matches the initialization pattern for the value. This code
    will probably need to be added back in though when issue #17198 is
    addressed.
 * this PR changes several `out` intent arguments to use `ref` instead
   because they are cases in which we would like to have type inference
   flow from the call site. These can be restored to `out` once issue
   #17198 is addressed.
 
Work on this PR revealed two issues present before this change that came
up in new ways due to the change:
 * problem with varargs and following arguments #17188
 * problem with split-init and ref vs. value tuples #17196

Future work:
 *  adjusting out formals with type query to infer type from call site
    #17198
 * improving the error message in test/arrays/intents/out/out-domain-no-elt.good
    to mention a generic type instead of no type
 
Reviewed by @vasslitvinov - thanks!

- [x] primers pass and do not leak
- [x] full local futures testing

### CODE REVIEW

**Comment by mppf** [comment]
These tests need to be updated

```
arrays/intents/out/complex-intents.chpl
classes/delete-free/undecorated-generic/cannot-default-init-c.chpl
classes/delete-free/undecorated-generic/cannot-default-init-cq.chpl
classes/nilability/demo2.chpl
classes/vass/ref-like-intents/out-subclass.chpl
classes/vass/ref-like-intents/out-superclass.chpl
domains/vass/domain-intents.chpl
functions/bradc/intents/intents-arrays.chpl
functions/bradc/intents/intents-arrays2.chpl
functions/bradc/intents/intents-arrays2a.chpl
functions/diten/wrappers/test_array_order_wrap.chpl
functions/jplevyak/intent-8-error.chpl
functions/vass/setter-passed-to-another-fun.chpl
functions/vass/varargs-1.chpl
library/packages/OrderedSet/general/lowerBound/osetLowerbound.chpl
library/packages/OrderedSet/general/upperBound/osetUpperbound.chpl
optimizations/constDomain/domainArgs.chpl
optimizations/constDomain/domainArgsConcrete.chpl
trivial/deitz/vars-scopes/infer1.chpl
types/integral/integralvariable2.chpl
types/partial/default-init-generic1.chpl
types/partial/default-init-generic2.chpl
types/records/bad-change-type/init-uses-default.chpl
types/records/expiring/more-deinit-points.chpl
types/records/ferguson/tuples/const1.chpl
types/records/intents/error-generic-out.chpl
types/records/intents/out-intent.chpl
types/records/intents/out-intent-error-array3.chpl
types/string/sungeun/c_string/intents.chpl
```

**Review by vasslitvinov — APPROVED** [review]
Wow a big change! Good stuff.

Be sure to update .good files containing "candidates are" after merging with master.

PR message:

Under "Besides the above changes, this PR makes a few other improvements:"
* "Such a type must be established later" -> "Such variable's type ..."
* remove the stray bullet

Perhaps explain briefly why special handling of out arguments with runtime types
can now be removed.

Mention that you changed a couple of `out` formals to `ref` formals because [...]

Code/tests:

In USR_PRINT text "the type for a generic out intent argument is inferred from the function body" :
* Add a dash: "out-intent"
* How about "in" instead of "from": "is inferred in the function body"

Also add a dash in "out intent formals do not participate in overload resolution" ?

In error text like "out formal 'a' is used in later formal 'b'"
I suggest "for a" instead : "out formal 'a' is used for a later formal 'b'"
Or perhaps "with a"?

Looking at `test/arrays/intents/out/out-domain-no-elt.good`
If possible, make the error message differentiate between
"is not initialized and has no type" vs. "is not initialized and has generic type"
where in this particular test it would be the latter.

See also other remarks in the code.

**Comment by mppf** [comment]
> Looking at test/arrays/intents/out/out-domain-no-elt.good
> If possible, make the error message differentiate between
> "is not initialized and has no type" vs. "is not initialized and has generic type"
> where in this particular test it would be the latter.

That is not as easy as one would hope. I will leave it for future work.

File: test/functions/intents/out/ambiguous-out-intent2.good [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ambiguous-out-intent2.chpl:9: In function 'main':
+ambiguous-out-intent2.chpl:11: error: ambiguous call 'f(int(64))'
+ambiguous-out-intent2.chpl:1: note: candidates are: f(out arg: int)
+ambiguous-out-intent2.chpl:5: note:                 f(out arg: real)
+ambiguous-out-intent2.chpl:11: note: out intent formals do not participate in overload resolution
```
[the remark this answers is not in the document]
Comment (mppf) [reply test/functions/intents/out/ambiguous-out-intent2.good:5]: candidates are is not printed on a line by itself in this .good file either. i'm not seeing any new issues after merging.

**Pull request merged**